drivers/stepper/adi/tmc5041_stepper_controller.c

```diff
@@ -0,0 +1,608 @@
+				    TMC5041_RAMPMODE_NEGATIVE_VELOCITY_MODE);
+		(void)tmc5041_write(config->controller, TMC5041_VMAX(config->index), velocity_fclk);
+		break;
+	}
+#ifdef CONFIG_STEPPER_ADI_TMC_STALLGUARD
+	(void)k_work_reschedule(&data->stallguard_dwork,
+				K_MSEC(CONFIG_STEPPER_ADI_TMC_STALLGUARD_ACTIVATION_PAUSE_IN_MSEC));
+#endif
+	return 0;
+}
+
+#ifdef CONFIG_STEPPER_ADI_TMC_RAMP_GEN
+
+int tmc5041_stepper_set_ramp(const struct device *dev,
+			     const struct tmc_ramp_generator_data *ramp_data)
+{
+	LOG_DBG("Stepper motor controller %s set ramp", dev->name);
+	const struct tmc5041_stepper_config *config = dev->config;
+	(void)tmc5041_write(config->controller, TMC5041_TZEROWAIT(config->index),
+			    ramp_data->tzerowait);
+	(void)tmc5041_write(config->controller, TMC5041_VSTART(config->index), ramp_data->vstart);
+	(void)tmc5041_write(config->controller, TMC5041_A1(config->index), ramp_data->a1);
+	(void)tmc5041_write(config->controller, TMC5041_AMAX(config->index), ramp_data->amax);
+	(void)tmc5041_write(config->controller, TMC5041_D1(config->index), ramp_data->d1);
+	(void)tmc5041_write(config->controller, TMC5041_DMAX(config->index), ramp_data->dmax);
+	(void)tmc5041_write(config->controller, TMC5041_V1(config->index), ramp_data->v1);
+	(void)tmc5041_write(config->controller, TMC5041_VMAX(config->index), ramp_data->vmax);
+	(void)tmc5041_write(config->controller, TMC5041_VSTOP(config->index), ramp_data->vstop);
+	(void)tmc5041_write(config->controller, TMC5041_VHIGH(config->index), ramp_data->vhigh);
+	(void)tmc5041_write(config->controller, TMC5041_VCOOLTHRS(config->index),
+			    ramp_data->vcoolthrs);
+	(void)tmc5041_write(config->controller, TMC5041_IHOLD_IRUN(config->index),
+			    ramp_data->iholdrun);
+	return 0;
+}
+
+#endif
+
+static int tmc5041_init(const struct device *dev)
+{
+	LOG_DBG("TMC5041 stepper motor controller %s initialized", dev->name);
+	struct tmc5041_data *data = dev->data;
+	const struct tmc5041_config *config = dev->config;
+
+	(void)k_mutex_init(&data->mutex);
+	if (!spi_is_ready_dt(&config->spi)) {
+		LOG_ERR("SPI bus is not ready");
+		return -ENODEV;
+	}
+
+	/* Read GSTAT register values to clear any errors SPI Datagram. */
+	uint32_t gstat_value;
+	(void)tmc5041_read(dev, TMC5041_GSTAT, &gstat_value);
+
+	/* Init non motor-index specific registers here. */
+	LOG_DBG("GCONF: %d", config->gconf);
+	(void)tmc5041_write(dev, TMC5041_GCONF, config->gconf);
+	LOG_DBG("Device %s initialized", dev->name);
+
+#ifdef CONFIG_STEPPER_ADI_TMC5041_INT
+	(void)activate_gpio_interrupt(dev);
+#endif
+	return 0;
+}
+
+static int tmc5041_stepper_init(const struct device *dev)
+{
+	const struct tmc5041_stepper_config *stepper_config = dev->config;
+	struct tmc5041_stepper_data *data = dev->data;
+
+	LOG_DBG("Controller: %s, Stepper: %s", stepper_config->controller->name, dev->name);
+#ifdef CONFIG_STEPPER_ADI_TMC_STALLGUARD
+	(void)tmc5041_write(stepper_config->controller, TMC5041_SWMODE(stepper_config->index), 0);
```

Comment:
Why activate stallguard, even before a move operation?

Reply:
I would activate stallguard before any kind of move operation. or?

Reply:
yes